api/chat.go

```diff
@@ -307,8 +309,45 @@ func (r chatRoutes) handleApprove(ctx tools.TUMLiveContext, msg []byte) {
 	broadcastStream(ctx.Stream.ID, broadcastBytes)
 }
 
+func (r chatRoutes) handleRetract(ctx tools.TUMLiveContext, msg []byte) {
+	var req wsIdReq
+	err := json.Unmarshal(msg, &req)
+	if err != nil {
+		log.WithError(err).Warn("could not unmarshal message retract request")
+		return
+	}
+	if ctx.User == nil || !ctx.User.IsAdminOfCourse(*ctx.Course) {
+		return
+	}
+
+	err = r.ChatDao.RetractChat(req.Id)
+	if err != nil {
+		log.WithError(err).Error("could not retract chat")
+		return
+	}
+
+	/* UserId should be the user who gets the message, to add dynamic user specific flags (e.g. Liked)
+	 * to the message payload. In this case the Message is freshly approved so no users should have interacted
+	 * with that message so far, so we pass 0 instead of a userId.
+	 */
+	chat, err := r.ChatDao.GetChat(req.Id, 0)
```

Comment:
I'm not really sure which userID I should set here. I don't necessarily want to re-implement the self-inflicted "DDOS attack", so could you please help me out @mono424? 😅

Reply:
If I understand right the message will be removed from all users so you could just use `0` here. The function only adds a `Liked` flag, which is individual for every user, but if the message disappears anyways this is probably not important.

But if you just want to remove the message maybe it's easier to just send an identifier of the message, in that case, you would not even need to call the `getChat()` function i guess :)
Than you could just do
```
broadcast := gin.H{
	"retract": req.Id,
}
```
or am I missing something ?

Reply:
Thanks! So for the meantime I will keep using `0` and maybe we can refactor this in an another PR.

Reply:
Can a moderator approve a message, than retract it and than approve it again? If so does the Likes be still present on the message? Because if yes I think there could a bug:
1. Message gets approved
2. UserA likes the messages
3. Message gets retracted
4. Message gets approved again
5. ! User does not see his like anymore

Is that an issue ? If not than nothing to worry about :)

Reply:
Yup, approving, retracting, and approving it again is possible. Likes are however still visible. Basically what retract does is set `visible` to false.

Reply:
Yes but the `Liked` flag will not be set right I believe. Because in the approve message function we also use `r.ChatDao.GetChat(req.Id, 0)` because we believe there are no likes on the post. If there are already likes on the post the `Liked` flag will be set not set for anyone.

Reply:
Yup. A possible (and probably the easiest one) solution would be to delete all likes of the message when retracting, right?

Reply:
yes I think that would be the best solution ! 👍

Reply:
edit: maybe not the best but the fastest to implement! 😄